File: 2017/ch3/ch3.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# vxf

import math

n = 368078
# n = 1024
# n = 12
# n = 23

# the squares always have an odd side length so we want to search only odd
# squares 2n+1
# we want to find the b right above n where (2b+1)**2 > n
b = math.floor((math.sqrt(n)-1)/2)

# and we also want to find the generating odd number
a = 2*b + 1

# we calculate the side length
m = 2 * (b + 1)

print(b + 1 + abs((n - (a*a)) % m - (b+1)))

File: 2017/ch3/ch3.erl
-module(ch3).
-export([main/0, main/1]).

solve(N) ->
  B = trunc((math:sqrt(N)-1)/2),
  A = 2*B + 1,
  M = 2 * (B + 1),
  B + 1 + abs((N - (A*A)) rem M - (B+1)).

main() ->
  main(368078).

main(N) ->
  io:write(solve(N)).
